Libraries/MLXVLM/Models/Paligemma.swift

```diff
@@ -0,0 +1,743 @@
+    }
+
+}
+
+// MARK: - Model
+
+private class PaliGemmaMultiModalProjector: Module, UnaryLayer {
+
+    @ModuleInfo var linear: Linear
+
+    public init(_ config: PaliGemmaConfiguration.VisionConfiguration) {
+        self.linear = Linear(config.hiddenSize, config.projectionDimensions, bias: true)
+    }
+
+    public func callAsFunction(_ x: MLXArray) -> MLXArray {
+        linear(x)
+    }
+}
+
+/// PaliGemma VLM
+///
+/// This is typically created by ``VLMModelFactory``.
+public class PaliGemma: Module, VLMModel, KVCacheDimensionProvider {
+
+    @ModuleInfo(key: "vision_tower") private var visionModel: Vision.VisionModel
+    @ModuleInfo(key: "language_model") private var languageModel: Language.LanguageModel
+    @ModuleInfo(key: "multi_modal_projector") private var multiModalProjector:
+        PaliGemmaMultiModalProjector
+
+    public let config: PaliGemmaConfiguration
+
+    public var vocabularySize: Int { config.vocabularySize }
+    public var kvHeads: [Int] { languageModel.kvHeads }
+    public var headDim: MLX.IntOrPair { languageModel.headDim }
+
+    public func loraLinearLayers() -> MLXLMCommon.LoRALinearLayers {
+        languageModel.model.layers.map { ($0.attention, ["q_proj", "v_proj"]) }
+    }
+
+    public init(_ config: PaliGemmaConfiguration) {
+        self.config = config
+        self._visionModel.wrappedValue = Vision.VisionModel(config.visionConfiguration)
+        self._languageModel.wrappedValue = Language.LanguageModel(config.textConfiguration)
+        self._multiModalProjector.wrappedValue = PaliGemmaMultiModalProjector(
+            config.visionConfiguration)
+    }
+
+    private func inputEmbeddings(inputIds: MLXArray, pixelValues: MLXArray?, mask: MLXArray) -> (
+        MLXArray, MLXArray
+    ) {
+        guard let pixelValues else {
+            return (inputIds, mask)
+        }
+
+        let inputEmbedding = languageModel.model.embedTokens(inputIds)
+        let (hiddenState, _, _) = self.visionModel(
+            pixelValues.transposed(0, 2, 3, 1).asType(inputEmbedding.dtype),
+            outputHiddenStates: true
+        )
+
+        var imageFeatures = hiddenState[.newAxis, .ellipsis].asType(pixelValues.dtype)
+        imageFeatures = multiModalProjector(imageFeatures)
+
+        return prepareInputsForMultimodal(
+            imageFeatures: imageFeatures, inputEmbedding: inputEmbedding,
+            inputIds: inputIds, attentionMask: mask)
+    }
+
+    private func prepareInputsForMultimodal(
+        imageFeatures: MLXArray, inputEmbedding: MLXArray, inputIds: MLXArray,
+        attentionMask: MLXArray
+    ) -> (MLXArray, MLXArray) {
+        let embedDimension = imageFeatures.dim(2)
+        let (batchSize, sequenceLength) = inputIds.shape2
+        var scaledImageFeatures = imageFeatures / pow(Float(config.hiddenSize), 0.5)
+        var finalEmbedding = zeros([batchSize, sequenceLength, embedDimension])
```

Comment:
The default data type of `zeros` is `fp32`. That will cause anything that works with this `finalEmbedding` to be upcasat to `fp32`.

Reply:
done